docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+# Incremental Generators
+
+## Summary
+
+Incremental generators are intended to be a new API that exists alongside
+[source generators](generators.md) to allow users to implement high performance
+incremental generation strategies.
+
+### High Level Design Goals
+
+- Allow for a finer grained approach to defining a generator
+- Scale source generators to support 'Roslyn/CoreCLR' scale projects in Visual Studio
+- Exploit caching between fine grained steps to reduce duplicate work
+- Support generating more items that just source texts
+- Exist alongside `ISourceGenerator` based implementations
+
+## Implementation
+
+An incremental generator is an implementation of `Microsoft.CodeAnalysis.IIncrementalGenerator`.
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface IIncrementalGenerator
+    {
+        void Initialize(IncrementalGeneratorInitializationContext context);
+    }
+}
+```
+
+As with source generators, incremental generators are defined in external
+assemblies and passed to the compiler via the `-analyzer:` option.
+Implementations are required to be annotated with the
+`Microsoft.CodeAnalysis.GeneratorAttribute` with an optional parameter
+indicating the languages the generator supports:
+
+```csharp
+[Generator(LanguageNames.CSharp)]
+public class MyGenerator : IIncrementalGenerator { ... }
+```
+
+An assembly can contain a mix of diagnostic analyzers, source generators and
+incremental generators.
+
+### Initialization
+
+`IIncrementalGenerator` has an `Initialize` method that is called by the host
+(either the IDE or the command-line compiler) exactly once. `Initialize` passes
+an instance of `IncrementalGeneratorInitializationContext` which can be used by
+the generator to register a set of callbacks that affect how future generation
+passes will occur.
+
+Currently `IncrementalGeneratorInitializationContext` supports two callbacks:
+
+- `RegisterForPostInitialization(GeneratorPostInitializationContext)`: the same
+  callback in shape and function as supported by Source Generators today
+- `RegisterExecutionPipeline(IncrementalGeneratorPipelineContext)`: replaces Execute, described below
+
+### Pipeline based execution
+
+Rather than a dedicated `Execute` method, an Incremental Generator instead
+creates an execution 'pipeline' as part of the initialization process via the
+`RegisterExecutionPipeline` method:
+
+```csharp
+public void Initialize(IncrementalGeneratorInitializationContext initContext)
```

Comment:
Above you called this value `context` not `initContext`. Don't care which is used but think it should be consistent within the document.